design/src/rfcs/rfc0045_configurable_serde.md

```diff
@@ -0,0 +1,212 @@
+let as_json = serde_json::to_string(&my_shape.serialize_ref(&settings));
+```
+
+### Customer Use Cases
+#### I want to embed a structure into my own types that implement `Serialize`
+The generated code includes two methods:
+`serialize_redacted` and `serialize_unredacted`.
+
+> Note: There is nothing in these implementations that rely on implementation details—Customers can implement these methods (or variants of them) themselves.
+
+These have the correct signatures to be used with `serialize_with`:
+```rust,ignore
+use generated_crate::serde::serialize_redacted;
+
+#[derive(Serialize)]
+struct MyStruct {
+    #[serde(serialize_with = "serialize_redacted")]
+    inner: SayHelloInput,
+}
+```
+#### I want to serialize data for testing
+This will be supported in the future. Currently `Deserialize` behavior is not covered by this RFC. Customers can take the same serialization settings they used.
+
+
+#### I want to dump a structured form of data into a database/logs/etc.
+This is possible by using the base APIs. If customers want to delegate another thread or piece of code to actually perform the serialization, they can use `.serialize_owned(..)` along with [erased-serde](https://crates.io/crates/erased-serde) to accomplish this.
+
+<!-- Explain the implementation of this new feature -->
+How to actually implement this RFC
+----------------------------------
+
+In order to provide configurable serialization, this defines the crate-local public trait `SerializeConfigured`:
+```rust,ignore
+/// Trait that allows configuring serialization
+/// **This trait should not be implemented directly!** Instead, `impl Serialize for ConfigurableSerdeRef<T>`
+pub trait SerializeConfigured {
+    /// Return a `Serialize` implementation for this object that owns the object.
+    ///
+    /// Use this if you need to create `Arc<dyn Serialize>` or similar.
+    fn serialize_owned(self, settings: SerializationSettings) -> impl Serialize;
+
+    /// Return a `Serialize` implementation for this object that borrows from the given object
+    fn serialize_ref<'a>(&'a self, settings: &'a SerializationSettings) -> impl Serialize + 'a;
+}
+```
+
+We also need to define `SerializationSettings`. The only setting currently exposed is `redact_sensitive_fields`:
+```rust,ignore
+
+#[non_exhaustive]
+#[derive(Copy, Clone, Debug, Default)]
+pub struct SerializationSettings {
+    /// Replace all sensitive fields with `<redacted>` during serialization
+    pub redact_sensitive_fields: bool,
+}
+```
+
+We MAY add additional configuration options in the future, but will keep the default behavior matching current behavior. Future options include:
+- Serialize `null` when a field is unset (the current default is to skip serializing that field)
+- Serialize blobs via a list of numbers instead of via base64 encoding
+- Change the default format for datetimes (current `HttpDate`)
+
+No objects actually implement `SerializeConfigured`. Instead, the crate defines two private structs:
```

Comment:
They are actually private structs allowing us to evolve the internal representation in the future

Reply:
Their fields are `pub(crate)`. Defining them in a separate common runtime crate wouldn't preclude us from evolving them in the future if we add `#[doc(hidden)]`. We have many examples of `#[doc(hidden)]` items in our runtime crates to make our lives easier and not have to generate them in every SDK / make them an `inlineable`.

Reply:
And the `SerializeConfigured` trait is public, and is what the user interacts with. That could also live in a separate common runtime crate.

Reply:
you can't have them both be centralized/public because you'll run into coherence issues, I believe. But we could have one of them be public.

Reply:
I think using `#[doc(hidden)]` isn't actually a great idea—you still need to follow semver since the generated code can rely on it so you can't actually use it to make breaking changes.